### ISSUE TITLE
Declare the signature of WSAAddressToString in _win32ifaces.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5430]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5430|
|Type|enhancement|
|Created|2011-12-11 04:56:35Z|
|Branch|https://github.com/twisted/twisted/tree/wsaatf-5430|

The `WSAAddressToStringA` function used by `win32GetLinkLocalIPv6Address` in `twisted/internet/test/_win32ifaces.py`, added for [#5084](https://github.com/twisted/twisted/issues/5084), does not have its signature declared.  ctypes very helpfully makes an (incorrect) guess, leading to an intermittent `WSAEFAULT` failure.

The 3rd argument, explicitly passed a null pointer (in the form of an integer 0), sometimes ends up as a non-null pointer in the underlying call.  ctypes presumably only manages to copy four bytes worth of zeros into a variable somewhere, and sometimes the _upper four bytes_ of that variable - since Win64 pointers are 8 bytes - contains non-zero garbage.

When this happens, the call typically fails, because the pointer is probably not to a mapped memory region (and even if it is, it would still lead to some wrong behavior).

Declaring the signature of `WSAAddressToString` will let ctypes know this is a pointer argument, which will let it fully initialize a null pointer, which should fix the `WSAEFAULT`.   This also removes the workaround added, allocating 16kB for `buf2`.  That value can be reduced to something like 64 bytes.

Attachments:

* [wsaaddresstostring-argtypes.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/12b/12b657fc6d7393ef00713442087235048dfd8a9a/6d4736b20544cab65007007ebe467b34416d261d.patch) (933 bytes) - added by exarkun on 2011-12-11 04:57:25Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5430 5430
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_wsaatf_5430 branches/wsaatf-5430
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__win64 win64
time__1323579395000000 1323579395000000
changetime__1323709994000000 1323709994000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

PenguinOfDoom actually tracked down this flaw and authored the attached patch.

There are no unit tests for this fix, as this kind of bogus memory initialization isn't something we have the tools to reliably unit test.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [33359]) Branching to 'wsaatf-5430'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

I took the liberty of producing some [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/wsaatf-5430).  Comments forthcoming.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Reviewin'.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

Thanks a lot, `exarkun` and `PenguinOfDoom`, for figuring this out.  It is pretty terrible.  I'm glad that this was ultimately our fault and not some kernel craziness.

Comments about the change itself:

    1. There really needs to be a link to the canonical reference for ''how'' that random little list of ctypes was derived.  I think [http://msdn.microsoft.com/en-us/library/ms741516(VS.85).aspx this is the right place]?
    2. Even with a link to the reference, the types might be a little mysterious.  I'd really like to have something I can eyeball and compare with the reference as mechanically as possible, then something that associates this with something else.  For example, something like:
```
#!py
LPSOCKADDR = c_void_p           # lpsaAddress
DWORD = c_int                   # dwAddressLength
# ...
WSAAddressToString.argtypes = [LPSOCKADDR, DWORD] # etc.
```
    Better yet get the actual arg names in there.
    3. I know the default `restype` is `c_int`, but why not set it anyway, just so it's explicit and nobody thinks we forgot it?
    4. `WSAIoctl` doesn't ahve _its_ `argtypes` declared either.  Looking at [the reference docs for the signature](http://msdn.microsoft.com/en-us/library/aa916335.aspx), I'm pretty sure I still see a couple of pointers being passed as 32-bit ints.  (Please, set its `restype` as well.)

As far as testing goes, I'm drawing a total blank.  I think that eyeballing this repeatedly is as good as we're likely to get.  I wish I had a better answer.

So I'd say the fix is ready to land if you see fit, as long as you address point 1.  If you feel like tackling point 2 too, go for it.

Now for some comments about the surrounding code, which you may want to address some other how:

    5. `WinDLL` is like an import.  Why is it happening at function scope rather than global scope?  If it actually does need to stay there, there should really be a comment as to why.  Presumably this module should never be imported on a non-Windows platform anyway.  (If these really do need to be module level, I'd actually really prefer it if you could just define another top-level function for each function so that we can hide away some of the `ctypes` crud; this function is long enough as it is and as we add more argtypes and restypes this is only going to get longer.)
    6. `WSAAddressToString failure` and `WSAIoctl failure` are pretty non-descriptive messages.  Can you at least get the error value in there?  Possibly the arguments that caused the error value?  I am anticipating that we'll really want to see those if we're ever fielding support requests for this code or anything related to it.
    7. While we're on those errors, wouldn't `WindowsError` be more appropriate?
    8. This API should eventually be public.  Interface enumeration is super helpful in some situations, and as this code demonstrates, super tricky.  Is there a ticket for that?
    9. In the comment about `WSAIoctl`: "which means there was not enough space in the buffer we have it."


Thanks again.  You can leave all of this latter feedback for later (except maybe the addition of `WSAIoctl.restypes` if it's not actually safe already).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Addressed points 1, 2, 3, 4, 5, and 9.  I'll file a ticket for 6, 7, and 8.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [33367]) Merge wsaatf-5430 

Author: PenguinOfDoom, exarkun
Reviewer: glyph
Fixes: [#5430](https://github.com/twisted/twisted/issues/5430)

Clean up the ctypes usage in the IPv6 link local interface lookup code for
Windows, fixing a bug on 64 bit Windows where a bad pointer will sometimes
be passed to `WSAAddressToString`.  Make some other changes to try to make
the code more maintainable in general, too.
